File: app/src/main/java/com/github/wheroj/goover2017_03_15/Book.java
package com.github.wheroj.goover2017_03_15;

import android.os.Parcel;
import android.os.Parcelable;

/**
 * Created by shopping on 2017/4/18 10:21.
 *
 * @description
 */

public class Book implements Parcelable {
    public String bookId;

    public String bookName;

    protected Book(Parcel in) {
        readFromParcel(in);
    }

    public Book(String bookId, String bookName) {
        this.bookId = bookId;
        this.bookName = bookName;
    }

    public static final Creator<Book> CREATOR = new Creator<Book>() {
        @Override
        public Book createFromParcel(Parcel in) {
            return new Book(in);
        }

        @Override
        public Book[] newArray(int size) {
            return new Book[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(bookId);
        dest.writeString(bookName);
    }

    //从Parcel中重建对象
    public void readFromParcel(Parcel in) {
        bookId = in.readString();
        bookName = in.readString();
    }
}

File: app/src/main/java/com/github/wheroj/goover2017_03_15/kotlin/Book.kt
package com.github.wheroj.goover2017_03_15.kotlin

import android.os.Parcel
import android.os.Parcelable

/**
 * Created by shopping on 2017/4/18 10:21.

 * @description
 */

class Book : Parcelable {
    var bookId: String = ""

    var bookName: String = ""

    protected constructor(`in`: Parcel) {
        readFromParcel(`in`)
    }

    constructor(bookId: String, bookName: String) {
        this.bookId = bookId
        this.bookName = bookName
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(bookId)
        dest.writeString(bookName)
    }

    //从Parcel中重建对象
    fun readFromParcel(`in`: Parcel) {
        bookId = `in`.readString()
        bookName = `in`.readString()
    }

    companion object {

        val CREATOR: Parcelable.Creator<Book> = object : Parcelable.Creator<Book> {
            override fun createFromParcel(`in`: Parcel): Book {
                return Book(`in`)
            }

            override fun newArray(size: Int): Array<Book?> {
                return arrayOfNulls(size)
            }
        }
    }
}
